apps/emqx_authn/src/emqx_authn_initializer.erl

```diff
@@ -0,0 +1,141 @@
+
+-include("emqx_authn.hrl").
+-include_lib("snabbkaffe/include/snabbkaffe.hrl").
+
+%% APIs
+-export([start_link/0,
+         start_link/1,
+         initialize/0,
+         deinitialize/0
+        ]).
+
+%% gen_statem callbacks
+-export([init/1,
+         callback_mode/0
+        ]).
+
+%% state functions
+-export([waiting_for_process/3,
+         initializing/3,
+         idle/3
+        ]).
+
+-type(opts() :: #{name := atom(),
+                  interval := non_neg_integer(),
+                  init_fun := fun((pid()) -> any())}).
+
+%%--------------------------------------------------------------------
+%% External interface
+%%--------------------------------------------------------------------
+
+-spec(start_link() -> gen_statem:start_ret()).
+start_link() ->
+        start_link(
+          #{name => ?AUTHN,
+            interval => ?DEFAULT_INTERVAL,
+            init_fun => fun(_) -> initialize() end}).
+
+-spec(start_link(opts()) -> gen_statem:start_ret()).
+start_link(#{} = Opts) ->
+        gen_statem:start_link(?MODULE, Opts, []).
+
+-spec(initialize() -> ok).
+initialize() ->
+    _ = ?AUTHN:register_providers(emqx_authn:providers()),
+
+    lists:foreach(
+      fun({ChainName, RawAuthConfigs}) ->
+              AuthConfig = emqx_authn:check_configs(RawAuthConfigs),
+              ?AUTHN:initialize_authentication(
+                 ChainName,
+                 AuthConfig)
+      end,
+      chain_configs()).
+
+-spec(deinitialize() -> ok).
+deinitialize() ->
+    ok = ?AUTHN:deregister_providers(provider_types()).
+
+%%--------------------------------------------------------------------
+%% gen_statem callbacks
+%%--------------------------------------------------------------------
+
+callback_mode() -> state_functions.
+
+init(#{init_fun := Fun,
+       name := Name,
+       interval := Interval} = Data)
+  when
+      is_function(Fun, 1)
+      and is_atom(Name)
+      and is_integer(Interval) ->
+    {ok, waiting_for_process, Data, [{state_timeout, 0, check_process}]}.
+
+waiting_for_process(state_timeout, check_process, #{name := Name,
+                                                   interval := Interval} = Data) ->
+    case whereis(Name) of
+        undefined ->
+            ?tp(info, emqx_authn_initializer_no_process, #{name => Name}),
+            {keep_state_and_data, [{state_timeout, Interval, check_process}]};
+        Pid ->
+            ?tp(info, emqx_authn_initializer_process_registered, #{name => Name}),
+            {next_state, initializing, Data, [{next_event, internal, {initialize, Pid}}]}
+    end.
+
+initializing(internal, {initialize, Pid}, #{init_fun := InitFun, name := Name} = Data) ->
+    _ = InitFun(Pid),
+    Ref = monitor(process, Pid),
+    ?tp(info, emqx_authn_initializer_process_initialized, #{name => Name}),
+    {next_state, idle, Data#{ref => Ref}, [{hibernate, true}]}.
+
+idle(info, {'DOWN', Ref, process, _Pid, _} = Msg, #{ref := Ref, name:= Name} = Data) ->
+    true = demonitor(Ref),
```

Comment:
It's not necessary, since the monitored process is dead

Reply:
✅